### PR TITLE
Add backup and restore functionality

### PR DESCRIPTION
**Pull request by mjhale**
Allows users to backup and restore the IndexedDB database via the `dexie-export-import` library.

When restoring, it will find the most recent app record within the database and load that version by replacing the current location with `/?r=VERSION`.

A form with buttons to create and restore a backup.

### CODE REVIEW

File: src/main/ogres/app/form/help.cljs [hunk 1]
```diff
@@ -3,18 +3,26 @@
...
+            [uix.core :refer [defui $ use-context use-ref use-state]]))
 
 (def ^:private confirm-upgrade
   "Upgrading will delete all your local data and restore this application to its original state.")
 
 (def ^:private confirm-delete
   "Delete all your local data and restore this application to its original state?")
 
+(def ^:private confirm-backup
+  "Backup your local state and assets?")
```
Comment (samcf) [inline src/main/ogres/app/form/help.cljs:15]:
```suggestion
(def ^:private confirm-backup
  "Backup your local data and images?")
```

File: src/main/ogres/app/form/help.cljs [hunk 2]
```diff
@@ -3,18 +3,26 @@
...
+
+(def ^:private confirm-restore
+  "Delete all your local data and restore this application to the provided backup?")
```
Comment (samcf) [inline src/main/ogres/app/form/help.cljs:18]:
```suggestion
(def ^:private confirm-restore
  "Delete all your local data and restore this application using the provided backup?")
```

File: src/main/ogres/app/form/help.cljs [hunk 3]
```diff
@@ -43,6 +51,57 @@
...
+            ($ :br)
+            ($ :p "Restoring a backup will "
+              ($ :strong "delete all your current local data") ". ")
+            ($ :br)
+            ($ :button.button.button-neutral
+              {:on-click
+               (fn []
+                 (if-let [_ (js/confirm confirm-backup)]
+                   (dispatch :storage/backup)))} "Create Backup")
+            ($ :br)
```
Comment (samcf) [inline src/main/ogres/app/form/help.cljs:68]: Just some styling changes.

```suggestion
            ($ :p {:style {:margin-bottom 4}}
              "Create a backup file that contains all your data and images. You
               can then use this file to restore your work on another computer
               or browser.")
            ($ :button.button.button-neutral
              {:on-click
               (fn []
                 (if-let [_ (js/confirm confirm-backup)]
                   (dispatch :storage/backup)))} "Create Backup")
            ($ :br)
            ($ :p {:style {:margin-bottom 4}}
              "Select a backup file to restore your data and images. Note that "
              ($ :strong "restoring from a file will delete all your current data") ".")
```

File: src/main/ogres/app/provider/storage.cljs [hunk 4]
```diff
@@ -2,6 +2,7 @@
   (:require [datascript.core :as ds]
             [datascript.transit :as dt]
             [dexie]
+            [dexie-export-import :refer [exportDB importDB]]
```
Comment (samcf) [inline src/main/ogres/app/provider/storage.cljs:5]:
```suggestion
            [dexie-export-import :refer [exportDB importDB peakImportFile]]
```

**Review by samcf — COMMENTED** [review]
Works perfectly! Its a little difficult to test if it correctly redirects to the last used release but I'll cut a release and test it out on production after its merged.

File: src/main/ogres/app/provider/storage.cljs [hunk 5]
```diff
@@ -102,11 +103,55 @@
...
+         (.then
+          (exportDB store)
+          (fn [blob]
+            (let [blob-url (js/window.URL.createObjectURL blob)
+                  shadow-anchor (js/document.createElement "a")]
+              (set! shadow-anchor -href blob-url)
+              (set! shadow-anchor -download
+                    (str "ogres-"
+                         (first (.split (.toISOString (js/Date.)) "T"))
+                         ".blob"))
```
[the remark this answers is not in the document]
Comment (mjhale) [reply src/main/ogres/app/provider/storage.cljs]: Using `.json` makes more sense as it's the format used for the exported file. Thanks!

**Comment by mjhale** [comment]
The suggestions have been committed, thanks for the help with cleaning things up!

**Review by samcf — APPROVED** [review]

**Pull request merged**